luigi/contrib/beam_dataflow.py

```diff
@@ -0,0 +1,451 @@
+        if self.max_num_workers:
+            output.append(f(self.dataflow_params.max_num_workers, self.max_num_workers))
+        if self.disk_size_gb:
+            output.append(f(self.dataflow_params.disk_size_gb, self.disk_size_gb))
+        if self.worker_machine_type:
+            output.append(f(self.dataflow_params.worker_machine_type, self.worker_machine_type))
+        if self.worker_disk_type:
+            output.append(f(self.dataflow_params.worker_disk_type, self.worker_disk_type))
+        if self.network:
+            output.append(f(self.dataflow_params.network, self.network))
+        if self.subnetwork:
+            output.append(f(self.dataflow_params.subnetwork, self.subnetwork))
+        if self.job_name:
+            output.append(f(self.dataflow_params.job_name, self.job_name))
+        if self.service_account:
+            output.append(f(self.dataflow_params.service_account, self.service_account))
+        if self.labels:
+            output.append(f(self.dataflow_params.labels, json.dumps(self.labels)))
+
+        return output
+
+    def _format_input_args(self):
+        job_input = self.input()
+        if isinstance(job_input, luigi.Target):
+            job_input = {"input": job_input}
+        elif not isinstance(job_input, dict):
+            raise ValueError("Input (requires()) must be dict type")
+
+        if not isinstance(self.file_pattern(), dict):
+            raise ValueError('file_pattern() must return a dict type')
+
+        input_args = []
+
+        for (name, targets) in job_input.items():
+            uri_targets = luigi.task.flatten(targets)
+            uris = [self._targets_to_uri_getter.get(
+                uri_target.__class__)(uri_target) for uri_target in uri_targets]
+            if isinstance(targets, dict):
+                """
+                If targets is a dict that means it had multiple outputs.
+                Make the input args in that case "<input key>-<task output key>"
+                """
+                names = ["%s-%s" % (name, key) for key in targets.keys()]
+            else:
+                names = [name] * len(uris)
+            for (arg_name, uri) in zip(names, uris):
+                pattern = self.file_pattern().get(name, 'part-*')
+                input_args.append(
+                    "--%s=%s" %
+                    (arg_name, uri.rstrip('/') + '/' + pattern)
+                )
+
+        return input_args
+
+    def _format_output_args(self):
+        job_output = self.output()
+        if isinstance(job_output, luigi.Target):
+            job_output = {"output": job_output}
+        elif not isinstance(job_output, dict):
+            raise ValueError("Input (requires()) must be dict type")
+
+        self.output_uris = {}
+        output_args = []
+
+        for (name, target) in job_output.items():
+            getter = self._targets_to_uri_getter.get(target.__class__)
+            uri = getter(target)
+
+            self.output_uris[name] = uri
```

Comment:
Is this dict really needed, apart from in tests? We are able to verify the uri values through the method return value `output_args`.

Reply:
yeah.... this is kind of a result of porting over needed features from our internal implementation, where we use them to publish output in the successful run callback :( @daikeshi also requested them for his team's [use case](https://github.com/spotify/spotify-tensorflow/blob/master/spotify_tensorflow/luigi/python_dataflow_task.py). I guess they could be re-derived in the implementation, though. what do you think?

Reply:
`self.output_uris` shadows the class variable, is that intended?

Reply:
it is intended--do you foresee an issue with that? even if that method is accidentally called twice it's idempotent so should end up with the same value.

Reply:
Why wouldn't an instance variable like this be good enough? What is the use case to have both class and instance variables?

Reply:
the original intent is to avoid having to recompute output URIs twice in any callback implementations but..... it's a pretty cheap computation tbh. Personally i'm ok taking it out. @daikeshi can you comment if you're ok with this? output uris can be recomputed in a one liner like `output_uris = dict(((name, self.get_target_path(target)) for (name, target) in self.output().items()))`

Reply:
ended up removing `self.output_uris` for the time being to move this PR forward.

Reply:
Sorry, somehow I missed this message.. yeah, this is totally fine! We can deal with it in our job class.